File: 2. Add Two Numbers/2. Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: Optional[ListNode]
        :type l2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        l1_val, l2_val, i = 0, 0, 0

        while(True) : 
            l1_val += (l1.val * (10 ** i))
            
            if (l1.next is None) :  
                break

            l1 = l1.next
            i = i + 1

        i = 0

        while(True) : 
            l2_val += (l2.val * (10** i))
            
            if (l2.next is None) :  
                break

            l2 = l2.next
            i = i + 1
        
        dummy = ListNode()
        current = dummy
        for value in [c for c in str(l1_val + l2_val)][::-1]:
            current.next = ListNode(int(value))
            current = current.next

        return dummy.next
        

File: 2. Add Two Numbers/2. Add Two Numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

#include <stdlib.h>

struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *createNode(int val)
{
    struct ListNode *node = (struct ListNode *)malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = NULL;
    return node;
}

struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2)
{
    struct ListNode dummy;
    struct ListNode *current = &dummy;
    dummy.next = NULL;

    int carry = 0;

    while (l1 || l2 || carry != 0)
    {
        int val1 = l1 ? l1->val : 0;
        int val2 = l2 ? l2->val : 0;

        int sum = val1 + val2 + carry;
        carry = sum / 10;
        int digit = sum % 10;

        current->next = createNode(digit);
        current = current->next;

        if (l1)
            l1 = l1->next;
        if (l2)
            l2 = l2->next;
    }

    return dummy.next;
}